std/conv.d

```diff
@@ -2816,374 +2828,113 @@ if (isInputRange!Source && isSomeChar!(ElementType!Source) && !is(Source == enum
+        if (isHex) p.popFront();
     }
 
     real ldval = 0.0;
-    char dot = 0;                        /* if decimal point has been seen */
+    char dot = 0;//, expIter;       /* dot is set if decimal point has been seen */
     int exp = 0;
-    long msdec = 0, lsdec = 0;
-    ulong msscale = 1;
+    ulong msdec = 0, lsdec = 0;
+    ulong msscale = 1;//, msscaleMax;
+    bool sawDigits;
+    //uint base;
+void parseDigits(bool ishex)(){
+    static if (ishex)
+    {
+        enum uint base = 16;
+        enum ulong msscaleMax = 0x1000_0000_0000_0000UL; // largest power of 16 a ulong holds
+        enum char expIter = 4; // iterate the base-2 exponent by 4 for every hex digit
+	alias checkDigit = isHexDigit;
+    }
+    else
+    {
+        enum uint base = 10;
+        enum ulong msscaleMax = 10_000_000_000_000_000_000UL; // largest power of 10 a ulong holds
+        enum char expIter = 1; // iterate the base-10 exponent once for every decimal digit
+	alias checkDigit = isDigit;
+    }
+
+    // Used to enforce that any mantissa digits are present
+    static if (ishex) sawDigits = false;
+    else sawDigits = startsWithZero;// && !isHex;
+
+    /*
+     * The following algorithm consists of mainly 2 steps (and maybe should
+     * be refactored into functions accordingly)
+     * 1) Parse the textual input into msdec/lsdec and exp variables:
+     *    Hex: input is 0xaaaaa...p+000... where aaaa is the mantissa in hex
+     *    and 000 is the exponent in decimal format with base 2.
+     *    Decimal: input is 0.00333...p+000... where 0.0033 is the mantissa
+     *    in decimal and 000 is the exponent in decimal format with base 10.
+     * 2) Convert msdec/lsdec and exp into native real format
+     */
 
-    if (isHex)
+    while (!p.empty)
     {
-        /*
-         * The following algorithm consists of mainly 3 steps (and maybe should
-         * be refactored into functions accordingly)
-         * 1) Parse the textual input into msdec and exp variables:
-         *    input is 0xaaaaa...p+000... where aaaa is the mantissa in hex and
-         *    000 is the exponent in decimal format.
-         * 2) Rounding, ...
-         * 3) Convert msdec and exp into native real format
-         */
+        int i = p.front;
+        while (checkDigit(i))
+        {
+            sawDigits = true;        /* must have at least 1 digit   */
 
-        int guard = 0;
-        // Used to enforce that any mantissa digits are present
-        bool anydigits = false;
-        // Number of mantissa digits (digit: base 16) we have processed,
-        // ignoring leading 0s
-        uint ndigits = 0;
+            /*
+             * convert letter to binary representation: First clear bit
+             * to convert lower space chars to upperspace, then -('A'-10)
+             * converts letter A to 10, letter B to 11, ...
+             */
+            i = isAlpha(i) ? ((i & ~0x20) - ('A' - 10)) : i - '0';
```

Comment:
This `isAlpha` check is unnecessarily done for the decimal digit parser, just wasn't sure how to statically change the way the conversion is done, aliased lambdas?